packages/vite/src/node/plugins/resolve.ts

```diff
@@ -545,90 +545,101 @@ export function resolvePackageEntry(
+            options
+          )
+          if (resolvedBrowserEntry) {
+            const content = fs.readFileSync(resolvedBrowserEntry, 'utf-8')
+            if (
+              (/typeof exports\s*==/.test(content) &&
+                /typeof module\s*==/.test(content)) ||
+              /module\.exports\s*=/.test(content)
+            ) {
+              // likely UMD or CJS(!!! e.g. firebase 7.x), prefer module
+              entryPoint = data.module
+            }
           }
+        } else {
+          entryPoint = browserEntry
         }
-      } else {
-        entryPoint = browserEntry
       }
     }
-  }
 
-  if (!entryPoint || entryPoint.endsWith('.mjs')) {
-    for (const field of options.mainFields || DEFAULT_MAIN_FIELDS) {
-      if (typeof data[field] === 'string') {
-        entryPoint = data[field]
-        break
+    if (!entryPoint || entryPoint.endsWith('.mjs')) {
+      for (const field of options.mainFields || DEFAULT_MAIN_FIELDS) {
+        if (typeof data[field] === 'string') {
+          entryPoint = data[field]
+          break
+        }
       }
     }
-  }
 
-  entryPoint = entryPoint || data.main || 'index.js'
+    entryPoint = entryPoint || data.main || 'index.js'
 
-  // resolve object browser field in package.json
-  const { browser: browserField } = data
-  if (targetWeb && isObject(browserField)) {
-    entryPoint = mapWithBrowserField(entryPoint, browserField) || entryPoint
-  }
+    // resolve object browser field in package.json
+    const { browser: browserField } = data
+    if (targetWeb && isObject(browserField)) {
+      entryPoint = mapWithBrowserField(entryPoint, browserField) || entryPoint
+    }
 
-  entryPoint = path.join(dir, entryPoint)
-  const resolvedEntryPoint = tryFsResolve(entryPoint, options)
+    entryPoint = path.join(dir, entryPoint)
+    const resolvedEntryPoint = tryFsResolve(entryPoint, options)
 
-  if (resolvedEntryPoint) {
-    isDebug &&
-      debug(
-        `[package entry] ${chalk.cyan(id)} -> ${chalk.dim(resolvedEntryPoint)}`
-      )
-    setResolvedCache('.', resolvedEntryPoint, targetWeb)
-    return resolvedEntryPoint
-  } else {
-    throw new Error(
-      `Failed to resolve entry for package "${id}". ` +
-        `The package may have incorrect main/module/exports specified in its package.json.`
-    )
+    if (resolvedEntryPoint) {
+      isDebug &&
+        debug(
+          `[package entry] ${chalk.cyan(id)} -> ${chalk.dim(
+            resolvedEntryPoint
+          )}`
+        )
+      setResolvedCache('.', resolvedEntryPoint, targetWeb)
+      return resolvedEntryPoint
+    } else {
+      packageEntryFailure(id)
+    }
+  } catch (e) {
+    packageEntryFailure(id, e.message)
   }
 }
 
+function packageEntryFailure(id: string, details?: string) {
+  throw new Error(
+    `Failed to resolve entry for package "${id}". ` +
+      `The package may have incorrect main/module/exports specified in its package.json` +
+      (details ? ': ' + details : '.')
```

Comment:
**suggestion (dx):** I prefer template literals due to they are better readable in such a situation

```suggestion
      (details ? `: ${details}` : '.')
```